src/test/lib/concurrency/stress_test.cpp

```diff
@@ -695,4 +707,87 @@ TEST_F(StressTest, VisibilityOfInsertsBeingRolledBack) {
   }
 }
 
+/**
+ * Test that adding and modifying a the TableKeyConstraints of a table concurrently does not lead to 
+ * deadlocks or inconsistencies (e.g. duplicate constraints).
+ */
+TEST_F(StressTest, AddModifyTableKeyConstraintsConcurrently) {
+  // Create a table with multiple TableKeyConstraints
+  auto table = std::make_shared<Table>(
+      TableColumnDefinitions{{"a", DataType::Int, false}, {"b", DataType::Int, false}, {"c", DataType::Int, false}},
+      TableType::Data, std::nullopt, UseMvcc::Yes);
+  table->add_soft_constraint(TableKeyConstraint{{ColumnID{0}}, KeyConstraintType::PRIMARY_KEY});
+
+  table->append({1, 1, 1});
+  table->append({2, 2, 2});
+  table->append({3, 3, 1});
+
+  Hyrise::get().storage_manager.add_table("dummy_table", table);
+
+  /** Run multiple modifications and additions concurrently that could potentially lead to race conditions:
+   * - `UccDiscoveryPlugin::_validate_ucc_candidates`
+   * - `StaticTableNode::unique_column_combinations`
+   * - `Table::soft_key_constraints().clear()` to force a revalidation of the constraints. This does not happen in
+   * production, but we want to test the concurrent insertion of constraints.
+   * NOTE: We do not execute `Table::add_soft_constraint` because we do not have the knowledge whether or not the
+   * constraint already exists.
+   */
+
+  Hyrise::get().default_pqp_cache = std::make_shared<SQLPhysicalPlanCache>();
+  Hyrise::get().default_lqp_cache = std::make_shared<SQLLogicalPlanCache>();
+  auto& pm = Hyrise::get().plugin_manager;
+  std::shared_mutex table_constraints_mutex;
+
+  pm.load_plugin(build_dylib_path("libhyriseUccDiscoveryPlugin"));
+  const auto validate_constraint = [&] {
+    std::shared_lock<std::shared_mutex> lock(table_constraints_mutex);
+    // Populate the plan cache.
+    const std::string sql = "SELECT b,c FROM dummy_table GROUP BY b,c;";
+    auto pipeline = SQLPipelineBuilder{sql}.create_pipeline();
+    pipeline.get_result_table();
+
+    pm.exec_user_function("hyriseUccDiscoveryPlugin", "DiscoverUCCs");
+  };
+
+  const auto clear_table_constraints = [&] {
+    std::unique_lock<std::shared_mutex> lock(table_constraints_mutex);
+    clear_soft_key_constraints(table);
+  };
+
+  const auto static_table_node_constraint_access = [&] {
+    std::shared_lock<std::shared_mutex> lock(table_constraints_mutex);
+    const auto static_table_node = StaticTableNode::make(table);
+    // Simply access the unique column combinations to require a `shared_lock`.
+    static_table_node->unique_column_combinations();
```

Comment:
Can you use a `StoredTableNode` here (because this is the one that used in actual queries)? Furthermore, I would almost bet on this code to be optimized away in release builds. Maybe wrap this in some sort of assert? E.g., like the final check, `ASSERT_LE(stored_table_node->unique_column_combinations().size(), 3);`?